### PR TITLE
Anton Ilinykh - UI Design Patterns Challenge 

### PR DESCRIPTION
**Pull request by ailinykh**

### CODE REVIEW

**Comment by caiozullo** [comment]
Hi @ailinykh, is this PR ready for review? 

If yes, please post a comment on the challenge page. Check the readme file for instructions. Thank you!

**Comment by ailinykh** [comment]
> Hi @ailinykh, is this PR ready for review?
> 
> If yes, please post a comment on the challenge page. Check the readme file for instructions. Thank you!

Oh sorry i've missed the error localization part. Will fix that

**Comment by caiozullo** [comment]
Hi @ailinykh, please post a comment on the academy's challenge page when this PR is ready for review. Check the readme file for instructions. Thank you!

**Comment by ailinykh** [comment]
Hi @caiozullo, I've done with comment, please check the academy's challenge page

File: MVP Module/Source/Feed Presentation/FeedErrorViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,12 @@
+//
+//  FeedErrorViewModel.swift
+//  UIDesignPatternsChallenge
+//
+//  Created by Anton Ilinykh on 22.02.2021.
+//  Copyright © 2021 Essential Developer. All rights reserved.
+//
+
+struct FeedErrorViewModel {
+	let message: String
```
Comment (caiozullo) [inline MVP Module/Source/Feed Presentation/FeedErrorViewModel.swift:10]: The `message` could be an optional `String?`, so the Presenter can decide when to show/hide the message.

For example, in the `didStartLoadingFeed()` method, the Presenter can call  `errorView.display(FeedErrorViewModel(message: .none))` to hide the error.

This way, the ViewController doesn't need to decide when to hide the error. It's easier to manage state when all changes are centralized in a single component.

File: MVP Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 2]
```diff
@@ -27,6 +28,7 @@ public final class FeedViewController: UITableViewController, UITableViewDataSou
 	
 	func display(_ viewModel: FeedLoadingViewModel) {
 		if viewModel.isLoading {
+			errorView.hideMessage()
```
Comment (caiozullo) [inline MVP Module/Source/Feed UI/Controllers/FeedViewController.swift:31]: The Presenter should decide when to show/hide messages. See my other comment.

File: MVP Module/Tests/Feed UI/FeedUIIntegrationTests.swift [hunk 3]
```diff
@@ -78,6 +78,19 @@ final class FeedUIIntegrationTests: XCTestCase {
...
+		let (sut, loader) = makeSUT()
+		
+		sut.loadViewIfNeeded()
+		XCTAssertEqual(sut.errorMessage, nil)
+		
+		loader.completeFeedLoadingWithError(at: 0)
+		XCTAssertEqual(sut.errorMessage, localized("FEED_VIEW_CONNECTION_ERROR"))
+		
+		sut.simulateTapOnErrorMessage()
+		XCTAssertEqual(sut.errorMessage, nil)
```
Comment (caiozullo) [inline MVP Module/Tests/Feed UI/FeedUIIntegrationTests.swift:91]: Missing assertion to check the error is hidden when the user reloads manually (e.g., pull to refresh).

File: MVVM Module/Tests/Feed UI/FeedUIIntegrationTests.swift [hunk 4]
```diff
@@ -78,6 +78,19 @@ final class FeedUIIntegrationTests: XCTestCase {
...
+		let (sut, loader) = makeSUT()
+		
+		sut.loadViewIfNeeded()
+		XCTAssertEqual(sut.errorMessage, nil)
+		
+		loader.completeFeedLoadingWithError(at: 0)
+		XCTAssertEqual(sut.errorMessage, localized("FEED_VIEW_CONNECTION_ERROR"))
+		
+		sut.simulateTapOnErrorMessage()
+		XCTAssertEqual(sut.errorMessage, nil)
```
Comment (caiozullo) [inline MVVM Module/Tests/Feed UI/FeedUIIntegrationTests.swift:91]: Missing assertion to check the error is hidden when the user reloads manually (e.g., pull to refresh).

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift [hunk 5]
```diff
@@ -20,12 +20,15 @@ final class FeedViewModel {
 	
 	var onLoadingStateChange: Observer<Bool>?
 	var onFeedLoad: Observer<[FeedImage]>?
+	var onError: Observer<String>?
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/FeedViewModel.swift:23]: `onError: Observer<String>?` could be `onErrorStateChange: Observer<String?>?`, so the ViewModel decides when to show/hide errors (instead of the view controller). It's easier to manage state when all changes are centralized in a single component.

For example, in the `loadFeed()` method, the ViewModel can call  `onErrorStateChange(.none)` to hide the error before loading the feed.

**Review by caiozullo — COMMENTED** [review]
Hi @ailinykh, you almost passed the challenge!

Please, check my review comments and let me know once you're done so I can review it again. Thank you 🙂

**Comment by ailinykh** [comment]
Hello @caiozullo, thanks for you feedback! Every single comment make a huge sense for me.
Please review the PR one more time

File: MVP Module/Source/Feed UI/Views/ErrorView.swift [hunk 6]
```diff
@@ -22,6 +22,14 @@ public final class ErrorView: UIView {
 		alpha = 0
 	}
 	
+	func display(_ model: FeedErrorViewModel) {
+		if let message = model.message {
+			show(message: message)
+		} else {
+			hideMessage()
+		}
+	}
```
Comment (caiozullo) [inline MVP Module/Source/Feed UI/Views/ErrorView.swift:31]: The view should ideally be decoupled from the presentable view model, so it can be reused with other models and in other projects. This logic should be in the controller.

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift [hunk 7]
```diff
@@ -20,12 +20,16 @@ final class FeedViewModel {
 	
 	var onLoadingStateChange: Observer<Bool>?
 	var onFeedLoad: Observer<[FeedImage]>?
+	var onError: Observer<String?>?
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/FeedViewModel.swift:23]: `onError` could be renamed to `onErrorStateChange` since it also represents the "no error" case with `.none`

**Review by caiozullo — COMMENTED** [review]
Hi @ailinykh,

Congratulations, you've passed the challenge 🥇

- All tests are passing ✅
- Your code is clean and handles edge cases ✅
- The UI and animations look and behave as expected ✅
- Light and Dark mode support ✅
- Added an extra language ✅

I left some review comments. Feel free to push changes.

Great job!

**Comment by caiozullo** [comment]
Nice one!
